ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -287,10 +299,90 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 			}
 			r.Log.Info("shutdownAfterJobFinishes set to true, we will delete cluster",
 				"RayJob", rayJobInstance.Name, "clusterName", fmt.Sprintf("%s/%s", rayJobInstance.Namespace, rayJobInstance.Status.RayClusterName))
-			return r.deleteCluster(ctx, rayJobInstance)
+			_, err = r.deleteCluster(ctx, rayJobInstance)
+			if err != nil && !errors.IsNotFound(err) {
+				return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+			}
+			_, err = r.deleteK8sJob(ctx, rayJobInstance)
+			if err != nil && !errors.IsNotFound(err) {
+				return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+			}
 		}
 	}
-	return ctrl.Result{}, nil
+	return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+}
+
+// getOrCreateK8sJob creates a Kubernetes Job for the Ray Job if it doesn't exist, otherwise return the existing one.
+func (r *RayJobReconciler) getOrCreateK8sJob(ctx context.Context, rayJobInstance *rayv1alpha1.RayJob) (string, error) {
+	jobName := rayJobInstance.Name
+	jobNamespace := rayJobInstance.Namespace
+
+	// Create a Job object with the specified name and namespace
+	job := &batchv1.Job{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: jobNamespace, Name: jobName}, job); err != nil {
+		if errors.IsNotFound(err) {
+			submitterTemplate, err := r.getSubmitterTemplate(rayJobInstance)
+			if err != nil {
+				return "", err
+			}
+			return r.createNewK8sJob(ctx, rayJobInstance, submitterTemplate)
+		}
+
+		// Some other error occurred while trying to get the Job
+		return "", err
+	}
+
+	// Job already exists, instead of returning an error we return a "success"
+	return jobName, nil
+}
+
+// getSubmitterTemplate builds the submitter pod template for the Ray job.
+func (r *RayJobReconciler) getSubmitterTemplate(rayJobInstance *rayv1alpha1.RayJob) (v1.PodTemplateSpec, error) {
+	var submitter_template v1.PodTemplateSpec
+
+	// Set the default value for the optional field SubmitterPodTemplate if not provided.
+	if rayJobInstance.Spec.SubmitterPodTemplate == nil {
+		submitter_template = common.GetDefaultSubmitterTemplate(rayJobInstance)
+	} else {
+		submitter_template = *rayJobInstance.Spec.SubmitterPodTemplate.DeepCopy()
+	}
+
+	// Set the command in the submitter pod template.
+	k8s_job_command, err := common.GetK8sJobCommand(rayJobInstance)
+	if err != nil {
+		return v1.PodTemplateSpec{}, err
+	}
+
+	submitter_template.Spec.Containers[0].Command = k8s_job_command
```

Comment:
1. If the command is specified, we should honor it.
2. If we want to hardcode `Containers[0]` as the submitter, we should make users know it (maybe add a log?).

Reply:
Good points, added here 6287936